rmw_connext_cpp/src/process_topic_and_service_names.cpp

```diff
@@ -27,116 +27,90 @@ bool
+    char * concat_str =
+      rcutils_format_string(allocator, "%s%s", ros_topic_prefix, topic_name);
+    if (!concat_str) {
+      RMW_SET_ERROR_MSG("could not allocate memory for topic string")
+      success = false;
+      goto end;
     }
-    *topic_str = DDS_String_dup(name_tokens.data[1]);
+    *topic_str = DDS_String_dup(concat_str);
+    allocator.deallocate(concat_str, allocator.state);
   } else {
-    RMW_SET_ERROR_MSG("incorrectly formatted topic name")
-    success = false;
+    *topic_str = DDS_String_dup(topic_name);
   }
 
 end:
-  // all necessary strings are copied into connext
-  // free that memory
-  if (rcutils_string_array_fini(&name_tokens) != RCUTILS_RET_OK) {
-    fprintf(stderr, "Failed to destroy the token string array\n");
-  }
   return success;
 }
 
 bool
 _process_service_name(
   const char * service_name,
   bool avoid_ros_namespace_conventions,
-  char ** service_str,
-  char ** request_partition_str,
-  char ** response_partition_str)
+  char ** request_topic_str,
+  char ** response_topic_str)
 {
   bool success = true;
-  rcutils_string_array_t name_tokens = rcutils_get_zero_initialized_string_array();
+
   rcutils_allocator_t allocator = rcutils_get_default_allocator();
 
-  if (rcutils_split_last(service_name, '/', allocator, &name_tokens) != RCUTILS_RET_OK) {
-    RMW_SET_ERROR_MSG(rcutils_get_error_string_safe())
-    success = false;
-    goto end;
-  }
-  if (name_tokens.size == 1) {
-    if (!avoid_ros_namespace_conventions) {
-      *request_partition_str = DDS_String_dup(ros_service_requester_prefix);
-      *response_partition_str = DDS_String_dup(ros_service_response_prefix);
-    }
-    *service_str = DDS_String_dup(name_tokens.data[0]);
-  } else if (name_tokens.size == 2) {
-    if (avoid_ros_namespace_conventions) {
-      // no ros_service_*_prefix, so store the user's namespace directly
-      *request_partition_str = DDS_String_dup(name_tokens.data[0]);
-      *response_partition_str = DDS_String_dup(name_tokens.data[0]);
-    } else {
-      // concat the ros_service_*_prefix with the user's namespace
-      char * request_concat_str = rcutils_format_string(
-        allocator,
-        "%s/%s", ros_service_requester_prefix, name_tokens.data[0]);
-      if (!request_concat_str) {
-        RMW_SET_ERROR_MSG("could not allocate memory for partition string")
-        success = false;
-        goto end;
-      }
-      char * response_concat_str = rcutils_format_string(
-        allocator,
-        "%s/%s", ros_service_response_prefix, name_tokens.data[0]);
-      if (!response_concat_str) {
-        allocator.deallocate(request_concat_str, allocator.state);
-        RMW_SET_ERROR_MSG("could not allocate memory for partition string")
-        success = false;
-        goto end;
-      }
-      *request_partition_str = DDS_String_dup(request_concat_str);
-      *response_partition_str = DDS_String_dup(response_concat_str);
+  if (!avoid_ros_namespace_conventions) {
```

Comment:
Consider to simply set the ros prefix to an empty string and run the same logic when concatenating the topic string. Because if I understand your code correctly, the only difference is whether you append the ros prefix or not, right?
This makes the logic for cleaning up the allocated strings simpler.


``` pseudo
string _ros_prefix = ros_prefix;
if (avoid_ros_namespace_conventions) {
   _ros_prefix = "";
}

topic = (%s%s%s, _ros_prefix, topic_name, "[Request|Reply]"
```